### ISSUE TITLE
use more cores to prerender more images

### ISSUE THREAD

**Issue by MPdaedalus** [BestImageViewer/geeqie#504]
I have issues rendering large images in a reasonable time. For example on my 4.5Ghz 4 core machine it takes 2 seconds to render a 70MP JPEG image and 3 seconds to render a 22MP JPEG 2000 image.

I noticed that geeqie will use at most two cores. One for thumbnailing and one for rendering/next Img render.  When flicking between images using scroll wheel geeqie locks up as it can not render the images in time and every scroll keeps the program locked for longer trying to catch up rendering. 

Would it be possible to change the preRender next image check box to a number so users could decide how many images to prerender. With all available cores used to preRender the images rather than just 1. The max value of "decoded image cache size" should be restricted to 80% of RAM rather than 1024MB so users like me that have 32GB ram can have lots of preRendered images to make scrolling between image quickly a much less annoying experience. 

Thanks for your time

**Comment by caclark** [comment]
Partial fix in commit 35f0119ce at geeqie.org.
Decoded image cache size increased.
There are no checks on your physical RAM size: if you set a value too high, you will end up using the swap file. And when that runs out...

**Comment by MPdaedalus** [comment]
Thanks for increasing the cache size, but will geeqie make use of that extra cache if it only prerenders the next image or is geeqie smart enough to keep pre rendering more images until the cache is full?

**Comment by caclark** [comment]
Unfortunately the cache is only used for files that have been already rendered. This is of use when you scroll back through images you have already viewed and rendered. The bigger the cache, the more images you can quickly scroll back through before Geeqie starts to re-render.

The problem of  Geeqie "freezing" when scrolling through images I will look at. Maybe that can be simply alleviated by putting something into the idle loop (I haven't noticed this effect, but I don't have 70MB images).

The problem of using multi-cores is more difficult, perhaps mostly because I don't know anything about this style of programming. The current design permits only a single image to be read-ahead. But the image loader software is thread-aware, so maybe there is hope that something can be done.

As a general comment, although there are other deficiencies Geeqie, the lack of real multi-core support is going to be the most difficult (for me, anyway) to fix.

**Comment by MPdaedalus** [comment]
As a java programmer I understand how doing multi-threaded programming right can be hard , I bet it's even harder in C. Maybe you could just let the single thread that does the read ahead render keep going instead of stopping after preRendering the next image. That's why I thought having an int value instead of a checkbox for "preload next image" would allow it to stop once it has hit X number of pre rendered images after current Image. Would be much easier than a multi-threaded setup.

In my first post I was referring to MegaPixels not megabytes. My images are only 1-4MB but are very high res. 5120x4300 for JPEG 2000 (3 second render time) and 10000x7100 for JPEG (1.5 second render time).  Turing off two pass and using nearest pixel render speeds it up slightly but its mostly decoding the image itself. For example in some 10K images I it goes from grey to green to colour as it renders each of the three channels.  The library you are using to render each jpeg is not multi threaded as it never uses more than one core. There are multithreaded jpeg decoding library's available> 
http://www.briancbecker.com/blog/2010/analysis-of-jpeg-decoding-speeds/

for jpeg2000 https://github.com/GrokImageCompression/grok seems to be fast.

hope this helps

**Comment by xsdg** [comment]
One fundamental issue here is that geeqie was originally designed and implemented as a single-main-threaded program.  There are places in the codebase that do things poorly (like stashing shared state in a global variable), and it's likely that there are others that I haven't run across yet.  So we'll need to do some code spelunking before we can make this happen.  It's definitely a good goal to have, but there are some places where we'll have to dig ourselves out of the past before we can move into the future (of course, let's not kid ourselves here.  "Future" really means "less recent past").

@mowgli it probably makes sense to reopen this bug just to keep track of the more fundamental request here.

**Comment by mowgli** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Hi Omari,

Am Di den  8. Aug 2017 um 20:10 schrieb Omari Stephens:
> One fundamental issue here is that geeqie was originally designed and implemented as a single-main-threaded program.  There are places in the codebase that do things poorly (like stashing shared state in a global variable), and it's likely that there are others that I haven't run across yet.  So we'll need to do some code spelunking before we can make this happen.  It's definitely a good goal to have, but there are some places where we'll have to dig ourselves out of the past before we can move into the future (of course, let's not kid ourselves here.  "Future" really means "less recent past").
> 
> @mowgli it probably makes sense to reopen this bug just to keep track of the more fundamental request here.

Sure.

Actually I am thinking about refactoring the code to archive some goals.
But that is something, not only one person can do and has to get
organized somehow.
- - Allow to unit test parts of the implementation by calling functions
  from test code
- - Allow to be more or less independent of the GUI framework. Partly to
  open the possibility to switch to another one (like QT).
- - And also to identify code pathes that can run in parallel.

However, I have no clue how to archive this and how long it will take.

At the moment I am developing a sort method to sort by colors.

Regards
   Klaus
- -- 
Klaus Ethgen                                       http://www.ethgen.ch/
pub  4096R/4E20AF1C 2011-05-16            Klaus Ethgen <Klaus@Ethgen.ch>
Fingerprint: 85D4 CA42 952C 949B 1753  62B3 79D0 B06F 4E20 AF1C
-----BEGIN PGP SIGNATURE-----
Comment: Charset: ISO-8859-1

iQGzBAEBCgAdFiEEMWF28vh4/UMJJLQEpnwKsYAZ9qwFAlmKLK8ACgkQpnwKsYAZ
9qzAJwv7B5FWUA92/IYiCV2q0ouUuPJwe53KzjdiVlEoZHSdb6zULPh3ZSf8Pu0Z
rJgR4pOcd1EVWtnHajQb8PKLn+pypzAh4EioCw/DCSn/PdZUe3+w1PE8cRai8xR/
XFl4kqeq7/8yGB4cjaYpcDAEXXcDJs77g2QhB0DliV0Drc5AwSS1P7GBcVBn3YOm
Z5flsUKNZ7IUzeOuG8j3TawoOD/rdjFb2GnTGXL59foaKW6wqvTzw6i2dpnvarlc
E3QJjcj/wLWu2TwLqSEA3V48noBJAKg1PNvwr+4iH2z8B9IVtS4geIFbo/DNJIL6
x7wDu0p7MnXKKVQJ2xpVCdruOd/wI2IKndcLUih6nb+uEHTWDQ8ZoQR0BwGpMA/s
M8Du5r/1rjFK28ZRlL+x+D31T5q3D27ftV7kgMCTRcm41EthjLRuaI9ykiIckiJF
qE/JWWdqk0jdufNr2WdR6Zmsxm7gZu2vQJL8v7vnvRJ1EnJNV7YdlrDlwgSijADT
wlLB1m8g
=6dqd
-----END PGP SIGNATURE-----

**Comment by caclark** [comment]
> When flicking between images using scroll wheel geeqie locks up as it can not render the images in time and every scroll keeps the program locked for longer trying to catch up rendering.

I assume this is a jpeg2000 problem. I can't find a libpixbufloader on my system to handle this format - which loader are you using?

**Comment by MPdaedalus** [comment]
i'm using geeqie 1.0 so what ever is in that is rendering jpeg 2000. The images were created with ImageMagick and have .jp2 extension. 

Its not just a jpeg 2000 issue (although it's the worst), I still get freezes when doing quick wheel scrolling with 10K JPEG images.

**Comment by caclark** [comment]
> I still get freezes when doing quick wheel scrolling with 10K JPEG images.

I'm sorry, but I am perplexed by this. I have re-compiled Geeqie v1.0 from sources, and I cannot replicate this problem. You may monitor Geeqie's actions by running from the command line:
geeqie --debug=1
As you scroll through images you will (or should) see in the terminal that Geeqie continues to output messages as it processes images.
Could you let me know which distribution and version you are using, and did you compile Geeqie from sources or use a package?

**Comment by MPdaedalus** [comment]
I found out that the pauses for 10K JPEG images were due to the fact that the images were on a spinning disk. When scrolling through images (whilst thumbnails are still loading for all the images) the disk could not keep up and geeqie paused as a result. When I moved the images to a SSD there were no pauses (although the grey>green>colour rendering for each image persists).

However some stalls for JPEG 2000 still happen (although not quite as bad) even when on a SSD.  For some reason the minor stalls only happen when fast wheel scrolling large 5K eg.5120x3700 images at 1:1 zoom level. If I fast scroll using "fit image to window" on the same images there are no pauses at all. This applies even when all the thumbnails have loaded.  The problem is noticeably worse with 10k JPEG 2000 images but fortunately I don't have many of them.
